numpy/_core/src/umath/ufunc_object.hpp

```diff
@@ -0,0 +1,292 @@
+#endif
+
+#ifndef NO_IMPORT_ARRAY
+    #define NO_IMPORT_ARRAY // supress warning: ‘int _import_array()’ defined but not used [-Wunused-function]
+#endif
+
+#ifndef _UMATHMODULE
+    #define _UMATHMODULE // initlize UFunc API for internal build
+#endif
+
+#include <numpy/ndarrayobject.h>
+#include <numpy/ufuncobject.h>
+
+#include "common.hpp"
+
+#include <functional>
+
+namespace np {
+
+namespace ufunc_helper {
+
+template <typename T>
+struct IsSliceBase_ : std::false_type {};
+template <typename T>
+struct IsSliceBase_<Slice<T>> : std::true_type {};
+// Determines if a type is a specialization of Slice.
+template <typename T, typename TBase=std::remove_cv_t<std::remove_reference_t<T>>>
+constexpr bool kIsSliceBase = IsSliceBase_<TBase>::value;
+
+template <typename T>
+struct FuncDeduce_;
+template <typename Ret, typename ..._Args>
+struct FuncDeduce_<std::function<Ret(_Args...)>> {
+    static constexpr int kNumArgs = sizeof...(_Args);
+    static constexpr int kNumIn = ((
+        std::is_const_v<std::remove_reference_t<_Args>> ||
+        !std::is_reference_v<_Args>
+    ) + ...);
+    static constexpr int kNumOut = kNumArgs - kNumIn;
+    static constexpr bool kIsSlice = (kIsSliceBase<_Args> && ...);
+    using Args = std::tuple<_Args...>;
+};
+// This metafunction deduces information about a callable object, including the number of input
+// and output arguments, whether any of the input arguments are slices, and the types of all the arguments.
+template<typename T>
+struct FuncDeduce : FuncDeduce_<decltype(std::function{std::declval<T>()})> {};
+// Determines the type ID of a Slice<T>'s element type.
+template <typename T, typename TBase=std::remove_cv_t<std::remove_reference_t<T>>>
+constexpr char ElementTypeID = static_cast<char>(kTypeID<typename TBase::ElementType>);
+
+} // namespace np::ufunc_helper
+
+/** Universal function object.
+ *
+ * This class simplifies the creation of UFunc objects by deducing
+ * the signature of function pointers in order to determine the desired operands of
+ * the overloaded functions (inner loops) rather than specify it manually similar to C API.
+ * Since this class still counts on the C API, the following limitations are observed:
+ *   - All C++ functions must have the same number of inputs and outputs
+ *   - Outputs parmaters must start at the end of the inputs.
+ *   - No support for return type.
+ *   - Only datatype Slice<T> is supported and there's no room for
+ *     supporting scalars yet.
+ *
+ * Some examples:
+ *
+ * @code
+ * template<typename T>
+ * static void Add(
+ *      // Prpamerts without references represent the inputs
+ *      np::Slice<T> in0, np::Slice<T> in1,
+ *      // and with reference `&` represents the output
+ *      np::Slice<T> &out
+ * )
+ * {
+ *     for (auto a = in0.begin(), b = in1.begin(), c = out.begin(),
+ *             end = out.end();  c != end; ++a, ++b, ++c) {
+ *         *c = *a + *b;
+ *     }
+ * }
+ *
+ * template<typename T>
```

Comment:
Yes but then I changed it to meta objects rather than function pointers due to @seberg concerns, I kept the doc as-is since we still haven't decided to favor one another.

Function pointers:
```C++
template<typename T>
void Add(np::Slice<T> in0, np::Slice<T> in1, np::Slice<T> &out)
{...}

UFuncObject add_obj(
    "add", "this doc of add function",
    Add<Float>, Add<Double>
);
```
Meta objects:
```C++
template <typename T>
struct AddOP {
    void operator ()(Slice<T> in0, Slice<T> in1, Slice<T> &out)
    {...}
};
UFuncObject<AddOP<Float>, AddOP<Double>> add_obj(
    "add", "this doc of add function"
);
```

Reply:
note that lambda can be supported too and also both styles above at the same time, however for simplicity of code I left only one.

Reply:
It would seem best to pick just one.  I tried to read the discussion with Stefan more on top, but it left me unclear what the advantage of function pointer vs meta object might be. Is it that one could quite easily use meta objects to wrap old-style ufunc loops?

Anyway, I think both styles will work, and I *really* like how either allows one to focus on what the loop does rather than on how the iteration works.

Reply:
I think the point is the meta object should be able to optimize things away (so that the pointer arrays don't need to be copied around on every call).  But I am not *sure* if I understood it right in the call.
This was indeed my main concern:  That we for-ever keep translating different function signature styles.  If this is all optimized away, then that doesn't matter much.

Reply:
> the advantage of function pointer vs meta object might be

Passing the function pointer as Sebastian explained will add an extra call for each time the inner loop is executed since the C++ function pointer will be stored on C ufunc data arbitrarily while the current approach (Meta-object function) is statically wrapping the C legacy function which allows the compiler to optimize it out.

Choosing function pointers provides more flexibility when it comes to sharing functions among different translation units which I would personally prefer also it improves the readability a bit, later on, we can cut extra calls by letting C++ take the lead over C code and initialize its own `PyType_Slot`.

Anyway, I'm still happy with meta-objects if this going to resolve this discussion and we can later discuss moving to C++ function pointers once we get rid of template sources and the umath python generator.

Reply:
Ah, thanks, both, that really helps!